pkg/server/api/svid/v1/service.go

```diff
@@ -99,10 +87,128 @@ func (s *service) MintJWTSVID(ctx context.Context, id spiffeid.ID, audience []st
 	return nil, status.Error(codes.Unimplemented, "not implemented")
 }
 
-func (s *service) NewX509SVID(ctx context.Context, entryID string, csr *x509.CertificateRequest) (*api.X509SVID, error) {
-	return nil, status.Error(codes.Unimplemented, "not implemented")
+func (s *service) BatchNewX509SVID(ctx context.Context, req []*BatchNewX509SVIDRequest) ([]*BatchNewX509SVIDResponse, error) {
+	// TODO: must I add metrics here? or middleware take care about it?
+	log := rpccontext.Logger(ctx)
+	callerID, ok := rpccontext.CallerID(ctx)
+	if !ok {
+		log.Error("CallerID is required")
+		return nil, status.Error(codes.Internal, "callerID is required")
+	}
+
+	regEntries, err := regentryutil.FetchRegistrationEntries(ctx, s.ds, callerID.String())
+	if err != nil {
+		log.WithError(err).Error("Failed to fetch registration entries")
+		return nil, status.Error(codes.Internal, "failed to fetch registration entries")
+	}
+
+	regEntriesMap := make(map[string]*common.RegistrationEntry)
+	for _, entry := range regEntries {
+		regEntriesMap[entry.EntryId] = entry
+	}
+
+	var resp []*BatchNewX509SVIDResponse
+	for _, eachReq := range req {
+		newSvid, err := s.newX509SVID(ctx, eachReq, regEntriesMap)
+		resp = append(resp, &BatchNewX509SVIDResponse{
+			Svid: newSvid,
+			Err:  err,
+		})
+	}
+	return resp, nil
+}
+
+func (s *service) newX509SVID(ctx context.Context, req *BatchNewX509SVIDRequest, regEntries map[string]*common.RegistrationEntry) (*api.X509SVID, error) {
+	log := rpccontext.Logger(ctx).WithField(telemetry.RegistrationID, req.EntryID)
+	entry, ok := regEntries[req.EntryID]
+	if !ok {
+		log.Error("Invalid registration entry: not allowed")
+		return nil, status.Errorf(codes.InvalidArgument, "invalid entry id: %q is not allowed", req.EntryID)
+	}
+
+	csr := req.Csr
+	spiffeID, err := s.validateCSR(csr, log)
```

Comment:
MintX509SVID and BatchNewX509SVID have different validation requirements for the CSR. Specifically, the MintX509SVID CSR has a few required fields but BatchNewX509SVID only uses/expects the public key.